src/memory.rs

```diff
@@ -121,7 +146,14 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
 
     // TODO(solson): Track which allocations were returned from __rust_allocate and report an error
     // when reallocating/deallocating any others.
-    pub fn reallocate(&mut self, ptr: Pointer, new_size: usize) -> EvalResult<'tcx, ()> {
+    pub fn reallocate(&mut self, ptr: Pointer, new_size: usize) -> EvalResult<'tcx, Pointer> {
+        if ptr.points_to_zst() {
+            if new_size != 0 {
```

Comment:
Oh, also, put the whole `if ptr.points_to_zst()` block after the below `if ptr.offset != 0` block.